pkg/syncer/cnsoperator/controller/csinodetopology/csinodetopology_controller.go

```diff
@@ -0,0 +1,336 @@
+		return err
+	}
+	// eventBroadcaster broadcasts events on csinodetopology instances to the event sink
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartRecordingToSink(
+		&typedcorev1.EventSinkImpl{
+			Interface: k8sclient.CoreV1().Events(""),
+		},
+	)
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: csinodetopologyv1alpha1.GroupName})
+	return add(mgr, newReconciler(mgr, clusterFlavor, configInfo, recorder))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager, clusterFlavor cnstypes.CnsClusterFlavor,
+	configInfo *cnsconfig.ConfigurationInfo, recorder record.EventRecorder) reconcile.Reconciler {
+	return &ReconcileCSINodeTopology{client: mgr.GetClient(), scheme: mgr.GetScheme(),
+		clusterFlavor: clusterFlavor, configInfo: configInfo, recorder: recorder}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	_, log := logger.GetNewContextWithLogger()
+
+	// Create a new controller
+	c, err := controller.New("csinodetopology-controller", mgr, controller.Options{Reconciler: r,
+		MaxConcurrentReconciles: defaultMaxWorkerThreadsForCSINodeTopology})
+	if err != nil {
+		log.Errorf("Failed to create new CSINodetopology controller with error: %+v", err)
+		return err
+	}
+
+	// Initialize backoff duration map
+	backOffDuration = make(map[string]time.Duration)
+
+	// Predicates are used to determine under which conditions
+	// the reconcile callback will be made for an instance.
+	pred := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return true
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			// The CO calls NodeGetInfo API just once during the node registration,
+			// therefore we do not support updates to the spec after the CR has been reconciled.
+			return false
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			// Instances are deleted by the garbage collector automatically after the corresponding NodeVM is deleted.
+			// No reconcile operations are required.
+			return false
+		},
+	}
+
+	// Watch for changes to primary resource CSINodeTopology
+	err = c.Watch(&source.Kind{Type: &csinodetopologyv1alpha1.CSINodeTopology{}}, &handler.EnqueueRequestForObject{}, pred)
+	if err != nil {
+		log.Errorf("Failed to watch for changes to CSINodeTopology resource with error: %+v", err)
+		return err
+	}
+	return nil
+}
+
+// blank assignment to verify that ReconcileCSINodeTopology implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileCSINodeTopology{}
+
+// ReconcileCSINodeTopology reconciles a CSINodeTopology object
+type ReconcileCSINodeTopology struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client        client.Client
+	scheme        *runtime.Scheme
+	clusterFlavor cnstypes.CnsClusterFlavor
```

Comment:
Why does clusterFlavor need to part of this struct?

Reply:
Good point, removed it.